[improve][txn] Allow superusers to abort transactions

nicoloboschi opened the pull request:
### Motivation

Super users must be always allowed to abort a transaction even if they're not the original owner.

### Modifications

* Check that only owner or superusers are allowed to perform txn operations (end, add partition and add subscription)

### Verifying this change

- [x] Make sure that the change passes the CI checks.


### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

michaeljmarshall commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2496,6 +2525,34 @@ protected void handleEndTxn(CommandEndTxn command) {
+        if (service.isAuthenticationEnabled() && service.isAuthorizationEnabled()) {
+            return getBrokerService()
+                    .getAuthorizationService()
+                    .isSuperUser(getPrincipal(), getAuthenticationData());
```
Does this work correctly for with proxy and original auth data? I will try to review more closely tomorrow.

michaeljmarshall replied:
One issue with this code is that we should always get this data from the netty event loop that the channel runs on.

michaeljmarshall commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2443,9 +2444,15 @@ protected void handleAddPartitionToTxn(CommandAddPartitionToTxn command) {
+                        return failedFutureTxnNotOwned(txnID);
+                    }
+                    return transactionMetadataStoreService
+                            .addProducedPartitionToTxn(txnID, command.getPartitionsList());
```
This is an unsafe call. We must copy the value because `command` is mutable.

michaeljmarshall commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2496,6 +2525,34 @@ protected void handleEndTxn(CommandEndTxn command) {
+                    if (service.isAuthenticationEnabled() && service.isAuthorizationEnabled()) {
+                        return getBrokerService()
+                                .getAuthorizationService()
+                                .isSuperUser(checkOwner, getAuthenticationData());
```
This is also unsafe. We should not call `getAuthenticationData()` from another thread. We can update the `thenCompose` to `thenComposeAsync` and run it on the `ctx.executor()`.

michaeljmarshall commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -2596,23 +2660,28 @@ protected void handleEndTxnOnSubscription(CommandEndTxnOnSubscription command) {
+                                return failedFutureTxnTcNotAllowed(txnID);
+                            }
+                            return subscription.endTxn(txnidMostBits, txnidLeastBits, txnAction, lowWaterMark);
+                        }).whenComplete((ignored, e) -> {
```
This is a nit, but I've been looking closely at the netty implementation, and I noticed that when we write to the channel, netty will schedule an event on the `ctx.executor()` if the calling thread is not the event loop (the executor). Would it make sense to prevent context switching and just schedule this to run on the ctx's event loop?

nicoloboschi commented:
@michaeljmarshall I've addressed your comments here: https://github.com/apache/pulsar/pull/19517
The last one is very interesting but I think we should do it for all the ServerCnx methods in one shot in another pull request.

michaeljmarshall commented:
> The last one is very interesting but I think we should do it for all the ServerCnx methods in one shot in another pull request.

I completely agree, thanks for addressing my feedback.
